kernel/arch/m68k/loader/Memory.cpp

```diff
@@ -0,0 +1,349 @@
+        physList.PushBack(reserved);
+
+        physList.Sort(MemoryBlockComparer{});
+    }
+
+    static void MapPage(uintptr_t vaddr, uintptr_t paddr, bool write)
+    {
+        const size_t pml3Idx = (vaddr >> 25) & 0x7F;
+        const size_t pml2Idx = (vaddr >> 18) & 0x7F;
+        const size_t pml1Idx = (vaddr >> 12) & 0x3F;
+
+        PageTable* pml3 = reinterpret_cast<PageTable*>(ptRoot);
+        PageTable* pml2 = nullptr;
+        if ((pml3->entries[pml3Idx] & PteResident) == 0)
+        {
+            uintptr_t page = AllocPages(1);
+            if (page == 0)
+                Panic(PanicReason::InternalAllocFailure);
+            sl::memset(sl::NativePtr(page).ptr, 0, sizeof(PageTable));
+
+            page |= PteResident;
+            pml3->entries[pml3Idx] = page;
+        }
+        pml2 = reinterpret_cast<PageTable*>(pml3->entries[pml3Idx] & ~0x1FF);
+
+        PageTable* pml1 = nullptr;
+        if ((pml2->entries[pml2Idx] & PteResident) == 0)
+        {
+            uintptr_t page = AllocPages(1);
+            if (page == 0)
+                Panic(PanicReason::InternalAllocFailure);
+            sl::memset(sl::NativePtr(page).ptr, 0, sizeof(PageTable));
+
+            page |= PteResident;
+            pml2->entries[pml2Idx] = page;
+        }
+        pml1 = reinterpret_cast<PageTable*>(pml2->entries[pml2Idx] & ~0xFF);
+
+        uint32_t* pte = &pml1->entries[pml1Idx];
+        *pte = paddr | PteResident | PteSupervisor | PteCopyback;
+        if (!write)
+            *pte |= PteWrite;
+    }
+
+    void InitMemoryManager()
+    {
+        freeHeadersUsed = 0;
+
+        //create a map of physical memory
+        sl::CNativePtr chunkScan = FindBootInfoTag(BootInfoType::MemChunk);
+        while (chunkScan.ptr != nullptr)
+        {
+            auto chunk = chunkScan.Offset(sizeof(BootInfoTag)).As<BootInfoMemChunk>();
+            uintptr_t base = chunk->addr;
+            size_t length = chunk->size;
+
+            if (base == 0)
+            {
+                base += PageSize;
+                length -= sl::Min(length, PageSize);
+            }
+
+            if (length != 0)
+            {
+                MemoryBlock* header = AllocMemoryBlock();
+                header->type = MemoryType::Usable;
+                header->base = base;
+                header->length = length;
+                physList.PushBack(header);
+                NPL_LOG("Usable memory chunk: 0x%x->0x%x (0x%x bytes).\r\n", base, length, length - base);
+            }
+
+            chunkScan = chunkScan.Offset(chunkScan.As<BootInfoTag>()->size);
+            chunkScan = FindBootInfoTag(BootInfoType::MemChunk, chunkScan);
+        }
+
+        const uintptr_t loaderBase = (uintptr_t)LOADER_BLOB_BEGIN;
+        const uintptr_t loaderTop = (uintptr_t)LOADER_BLOB_END;
+        ReservePhysicalRange(loaderBase, loaderTop, MemoryType::Reclaimable);
+
+        if (auto initrdPtr = FindBootInfoTag(BootInfoType::InitRd); initrdPtr.ptr != nullptr)
```

Comment:
is that correct? you initialize a variable within an if statement? (i assume it's probably a c++ syntax)

Reply:
Yeah it's allowed, not a commonly used feature, but it limits the variable (`initrdPtr`) to the inside of the if statement.